File: src/main/java/com/example/today_i_baekjoon/service/CommitService.java
package com.example.today_i_baekjoon.service;

import com.example.today_i_baekjoon.domain.Commit;
import com.example.today_i_baekjoon.domain.User;
import com.example.today_i_baekjoon.dto.CommitInfo;
import com.example.today_i_baekjoon.dto.CommitWebhookRequest;
import com.example.today_i_baekjoon.repository.CommitRepository;
import com.example.today_i_baekjoon.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class CommitService {

    private final UserRepository userRepository;
    private final CommitRepository commitRepository;
    private final Pattern pattern = Pattern.compile("^\\[(?<rank>.*)] Title: (?<title>.*), Time:.*");

    public void addCommits(User user, List<CommitWebhookRequest.Commit> commitRequests) {
        List<Commit> commits = commitRequests
                .stream()
                .filter(commitRequest -> !commitRepository.existsByCommitUrl(commitRequest.getUrl()))
                .filter(commitRequest -> pattern.matcher(commitRequest.getMessage()).matches())
                .map(commitRequest -> {
                    Matcher matcher = pattern.matcher(commitRequest.getMessage());
                    String rank = "", title = "";
                    if (matcher.matches()) {
                        rank = matcher.group("rank");
                        title = matcher.group("title");
                    }
                    
                    return Commit.builder()
                            .user(user)
                            .commitDate(commitRequest.getTimestamp())
                            .commitUrl(commitRequest.getUrl())
                            .problemTitle(title)
                            .problemRank(rank)
                            .build();
                })
                .toList();

        commitRepository.saveAll(commits);
    }
    
    public List<CommitInfo> getAllCommitsByDate(LocalDate date) {
        return commitRepository
                .findAllByCommitDate(date)
                .stream()
                .map(CommitInfo::fromEntity)
                .collect(Collectors.toList());
    }
    
    public List<CommitInfo> getAllCommitsBetweenDate(LocalDate start, LocalDate end) {
        return commitRepository
                .findAllByCommitDateBetween(start, end)
                .stream()
                .map(CommitInfo::fromEntity)
                .collect(Collectors.toList());
    }
    
    public List<CommitInfo> getAllCommitsBetweenDate(LocalDate start, LocalDate end, User user) {
        return commitRepository
                .findAllByCommitDateBetweenAndUser(start, end, user)
                .stream()
                .map(CommitInfo::fromEntity)
                .collect(Collectors.toList());
    }
    
    public Map<User, List<Boolean>> getSolvedListBetweenDate(LocalDate start, LocalDate end) {
        List<User> users = userRepository.findAll();
        List<Commit> commits = commitRepository.findAllByCommitDateBetween(start, end);
        
        Map<User, List<Boolean>> result = new HashMap<>();
        
        users.forEach(user -> result.put(user, new ArrayList<>()));
        
        for (LocalDate date = start; !date.isAfter(end); date = date.plusDays(1)) {
            LocalDate finalDate = date;
            
            for(User user : result.keySet()) {
                Optional<Commit> solved = commits
                        .stream()
                        .filter(commit ->
                                commit.getCommitDate().isEqual(finalDate) &&
                                        commit.getUser().getUsername().equals(user.getUsername())
                        ).findAny();

                result.get(user).add(solved.isPresent());
            }
        }
        
        return result;
    }
}

File: src/main/kotlin/com/gidskql6671/today_i_baekjoon/service/CommitService.kt
package com.gidskql6671.today_i_baekjoon.service

import com.gidskql6671.today_i_baekjoon.domain.Commit
import com.gidskql6671.today_i_baekjoon.domain.User
import com.gidskql6671.today_i_baekjoon.dto.CommitInfo
import com.gidskql6671.today_i_baekjoon.dto.CommitWebhookRequest
import com.gidskql6671.today_i_baekjoon.dto.commitInfoFromEntity
import com.gidskql6671.today_i_baekjoon.repository.CommitRepository
import com.gidskql6671.today_i_baekjoon.repository.UserRepository
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.util.function.Consumer
import java.util.regex.Pattern
import java.util.stream.Collectors

@Service
class CommitService(
    private val commitRepository: CommitRepository,
    private val userRepository: UserRepository
) {
    private val pattern = Pattern.compile("^\\[(?<rank>.*)] Title: (?<title>.*), Time:.*")

    fun addCommits(user: User, commitRequests: List<CommitWebhookRequest.Commit>) {
        val commits = commitRequests
            .stream()
            .filter { !commitRepository.existsByCommitUrl(it.url) }
            .filter { pattern.matcher(it.message).matches() }
            .map {
                var rank = ""
                var title = ""

                val matcher = pattern.matcher(it.message)
                if (matcher.matches()) {
                    rank = matcher.group("rank")
                    title = matcher.group("title")
                }

                Commit(
                    user = user,
                    commitDate = it.timestamp.toLocalDate(),
                    commitUrl = it.url,
                    problemTitle = title,
                    problemRank = rank
                )
            }
            .toList()

        commitRepository.saveAll(commits)
    }

    fun getAllCommitsByDate(date: LocalDate): List<CommitInfo> {
        return commitRepository
            .findAllByCommitDate(date)
            .stream()
            .map { commitInfoFromEntity(it) }
            .collect(Collectors.toList())
    }

    fun getAllCommitsBetweenDate(start: LocalDate, end: LocalDate): List<CommitInfo> {
        return commitRepository
            .findAllByCommitDateBetween(start, end)
            .stream()
            .map { commitInfoFromEntity(it) }
            .collect(Collectors.toList())
    }

    fun getAllCommitsBetweenDate(start: LocalDate, end: LocalDate, user: User): List<CommitInfo> {
        return commitRepository
            .findAllByCommitDateBetweenAndUser(start, end, user)
            .stream()
            .map { commitInfoFromEntity(it) }
            .collect(Collectors.toList())
    }

    fun getSolvedListBetweenDate(start: LocalDate, end: LocalDate): Map<User, MutableList<Boolean>> {
        val users = userRepository.findAll()
        val commits = commitRepository.findAllByCommitDateBetween(start, end)
        val result: MutableMap<User, MutableList<Boolean>> = HashMap()

        users.forEach(Consumer { user: User -> result[user] = ArrayList() })

        var date = start
        while (!date.isAfter(end)) {
            val finalDate = date

            for (user in result.keys) {
                val solved = commits
                    .stream()
                    .filter { it.commitDate.isEqual(finalDate) && it.user.username == user.username }
                    .findAny()

                result[user]!!.add(solved.isPresent)
            }

            date = date.plusDays(1)
        }
        return result
    }
}